### ISSUE TITLE
Instance arguments and multi-sorted algebras

### ISSUE THREAD

**Issue by anuyts** [agda/agda#2670]
Instance arguments do not currently play well with certain multi-sorted algebras (as I discovered while implementing categories with families). Below is a code example in the simpler setting of reflexive graphs.

The example defines a type `IsRG (Node : Set) (Edge : Set) : Set`. The problem is the following: when Agda is in search of an instance of type `IsRG N E` (where `N` and `E` contain metavariables) and there is a unique solution to the problem "find types `N'` and `E'` and an instance `i : IsRG N' E'`", then Agda is not prepared to solve the metavariables in `N` and `E` to make both types match. Rather, it observes that both types *may not* match and gives up.

This is unproblematic in single-sorted algebras. However, when working with e.g. reflexive graphs, it means that if we have a computation from which only the type of nodes can be inferred, then Agda requires us to specify the relevant type of edges before it is confident that we are using the *only* reflexive graph with that set of nodes (and indeed the only reflexive graph whatsoever) in scope.

In CwFs, it means that we must make sure that the types `Ctx`, `Sub`, `Ty` and `Tm` can all be inferred from every computation. This can be partly remedied by introducing sorts one at a time (e.g. define a Category with Types but not Terms), but even then, one would like to not have to remind Agda of the definition of a context, when doing a type-level substitution `T [ σ ]`

In the code below, you can see that I had to add so many hints that it would currently seem easier not to use instance arguments at all. *Side issue: Also, I needed an additional hint when working with a record type of reflexive graphs; this is surprising behaviour and possibly [but probably not: see below] a bug.*

In short, I suggest that Agda should go all in when there is a unique solution; this includes solving metavariables to make that solution work.

```agda
module InstanceMultisort where

open import Relation.Binary.PropositionalEquality
open import Function

record IsRG (Node : Set) (Edge : Set) : Set where
  field
    src : Edge → Node
    tgt : Edge → Node
    rfl : Node → Edge
    eq-src-rfl : ∀{x} → src (rfl x) ≡ x
    eq-tgt-rfl : ∀{x} → tgt (rfl x) ≡ x
open IsRG {{...}} public

module norecord where

  source : ∀{Node Edge : Set} {{i : IsRG Node Edge}} → Edge → Node
  source x = src x

  lemma1 : ∀{Node Edge : Set} {{i : IsRG Node Edge}} → (e : Edge) → (Node ∋ source e) ≡ src e
  lemma1 e = refl

  rfl-src : ∀{Node Edge : Set} {{i : IsRG Node Edge}} → Edge → Edge
  rfl-src {Node} e = rfl (Node ∋ src e)

  lemma2 : ∀{Node Edge : Set} {{i : IsRG Node Edge}} → (n : Node) → rfl-src {Node} (rfl n) ≡ (Edge ∋ rfl n)
  lemma2 {Node}{Edge} n = cong (rfl {Node = Node}) (eq-src-rfl {Edge = Edge})

module yesrecord where

  record RG : Set₁ where
    constructor mkRG
    field
      Node : Set
      Edge : Set
      {{isRG}} : IsRG Node Edge
  open RG public

  source : ∀{rg} → Edge rg → Node rg
  source x = src x

  lemma1 : ∀{rg} → (e : Edge rg) → (Node rg ∋ source {mkRG (Node rg) (Edge rg)} e) ≡ src e
  --causes problems:
  --lemma1 : ∀{rg} → (e : Edge rg) → (Node rg ∋ source {_} e) ≡ src e
  lemma1 e = refl

  rfl-src : ∀{rg} → Edge rg → Edge rg
  rfl-src {rg} e = rfl (Node rg ∋ src e)

  lemma2 : ∀{rg} → (n : Node rg) → rfl-src {rg} (rfl n) ≡ (Edge rg ∋ rfl n)
  lemma2 {rg} n = cong (rfl {Node = Node rg}) (eq-src-rfl {Edge = Edge rg})
```

**Comment by m0davis** [comment]
In `yesrecord`, this works:
```agda
lemma1 : ∀{rg} → (e : Edge rg) → (Node rg ∋ norecord.source e) ≡ src e
```
So, you don't need require `{{i : IsRG Node Edge}}` as you did in `norecord`. Does that help?

**Comment by anuyts** [comment]
Well, what I think is surprising is that Agda does not expand the metavariable of type `RG` to two variables of type `Set` (which it can solve to `Node` and `Edge`) and an instance problem of type `IsRG Node Edge` which has a unique solution `isRG rg`. Your solution is a workaround (which boils down to not using the `RG` type), as is explicitly mentioning `mkRG (Node rg) (Edge rg)` as I did above, but I wouldn't expect this to be necessary.

**Comment by m0davis** [comment]
I think Agda is treating `source {_}` as `source {mkRG _ _ {{_}}}` rather than as you expect, `source {mkRG _ _ {{`[it](https://github.com/UlfNorell/agda-prelude/blob/3f99b98394f20d300249976eb1f5c24a9a006637/src/Prelude/Function.agda#L58)`}}`. That is, she's not invoking instance search in unification. (I'm not exactly sure why that is.) Perhaps your request is that she does so?

**Comment by m0davis** [comment]
To see why I think that, you could change `mkRG (Node rg) (Edge rg)` to `mkRG _ _` and Agda would work as you'd expect, rendering it (w/o error) as `source {mkRG _ _ {{it}}}`. Or try `mkRG _ _ {{_}}` and Agda gives the same error as if you had just `_`.

**Comment by m0davis** [comment]
Here's another workaround that requires yet one fewer hint (i.e. no need for `Node rg ∋`). It works because the below implicit argument to `isource` is solved with instance search, which recursively solves ```isRG``` via instance search. [That last statement is wrong. See my next comment.]
```agda
isource : ∀ {{rg}} → Edge rg → Node rg
isource x = src x

lemma1 : ∀ {{rg}} → (e : Edge rg) → isource e ≡ src e
```
Alternatively, if you'd like to keep the same interface to `lemma1`:
```
lemma1 : ∀ {rg} (let instance _ = rg) → (e : Edge rg) → isource e ≡ src e
```

**Comment by m0davis** [comment]
Sorry, no, the reason I gave for it working is not correct: it works b/c `rg` is the only instance of `RG` found.

**Comment by anuyts** [comment]
> I think Agda is treating `source {_}` as `source {mkRG _ _ {{_}}}`

Thanks for pointing out what happens there more clearly. Then that is probably not a bug.

> Perhaps your request is that she does so?

That could be another request, I guess. Although it is less clear to me what is the ideal situation here. One could argue that users do not expect instance search to happen when they are not calling a function with instance arguments anywhere.

But my main point remains the issue with multisorted algebras: that instance solutions are not accepted if the instance problem contains metavariables and the solution does not.

**Comment by jespercockx** [comment]
I think it is a conscious design decision of Agda to not solve metas during instance search, see e.g. https://github.com/agda/agda/issues/1272. If Agda would do so, it could later lead to an error during typechecking where you (the user) would expect an unresolved instance.

**Comment by UlfNorell** [comment]
There's also an issue with potential non-termination if you allow instantiation of metavariables. There are some details [here](http://agda.readthedocs.io/en/latest/language/instance-arguments.html#instance-resolution).

**Comment by Blaisorblade** [comment]
Also relevant: #1732, opened after Agda introduced this restriction. There @UlfNorell suggested functional dependencies as an alternative, while I mentioned that Lean claims to behave as we'd like. It's unclear (to me at least) whether their implementation approach could apply to Agda or not:

https://github.com/agda/agda/issues/1732#issuecomment-224003859

**Comment by anuyts** [comment]
The following seems an acceptable workaround, as many multisorted algebras have a 'base' sort:

```agda
module InstanceMultisort2 where

open import Relation.Binary.PropositionalEquality
open import Function

record IsRG (Node : Set) : Set₁ where
  field
    Edge : Set
    src : Edge → Node
    tgt : Edge → Node
    rfl : Node → Edge
    eq-src-rfl : ∀{x} → src (rfl x) ≡ x
    eq-tgt-rfl : ∀{x} → tgt (rfl x) ≡ x
open IsRG {{...}}

module norecord where

  source : ∀{Node : Set} {{i : IsRG Node}} → _ → Node
  source {Node} x = src x
  
  source' : ∀{Node : Set} {{i : IsRG Node}} → Edge {Node} → Node
  source' {Node} x = src x

  lemma1 : ∀{Node : Set} {{i : IsRG Node}} → (e : _) → source {Node} e ≡ src e
  lemma1 e = refl

  lemma1' : ∀{Node : Set} {{i : IsRG Node}} → (e : Edge {Node}) → source {Node} e ≡ src e
  lemma1' e = refl

  rfl-src : ∀{Node : Set} {{i : IsRG Node}} → _ → _
  rfl-src {Node} e = rfl {Node} (src e)

  rfl-src' : ∀{Node : Set} {{i : IsRG Node}} → Edge {Node} → Edge {Node}
  rfl-src' {Node} e = rfl {Node} (src e)

  lemma2 : ∀{Node : Set} {{i : IsRG Node}} → (n : Node) → rfl-src {Node} (rfl n) ≡ (rfl n)
  lemma2 {Node} n = cong (rfl {Node = Node}) (eq-src-rfl)
```

Remarkably, in each of the cases where i wrote an `_`, replacing it with `Edge` (without specifying `{Node}`) yields an instance search failure with a single candidate:

```
Failed to solve the following constraints:
  Resolve instance argument
    _r_35
      : {Node = Node₁ : Set} ⦃ i = i₁ : IsRG Node₁ ⦄ → IsRG _Node_34
  Candidates i : IsRG Node
  [43] IsRG.Edge _r_35 =< IsRG.Edge i : Set
  _44 := λ {Node} ⦃ i ⦄ e → e [blocked on problem 43]
  _40 := λ {Node} ⦃ i ⦄ e → e [blocked on problem 39]
  [39] IsRG.Edge _r_35 =< IsRG.Edge i : Set
```

**Comment by jespercockx** [comment]
This is probably a duplicate of #2760 (or vice versa)

**Comment by anuyts** [comment]
See #3419
